text/0001-remove_ambiguous_grammar.md

```diff
@@ -0,0 +1,244 @@
+
+## Introduce strand prefixes
+
+The only way can currently be distinguished is if one strand fails to parse as any of the other types. This can lead to problems where a use wanted a certain value to be a plain strand but it happened to match another type of strand and is thus converted into that specific type.
+
+And example is `"5:00"` which brought up in an issue. The user wanted to store this value as a plain strand but it happened to match a thing strand.
+
+I propose we introduce specific strand prefixes for specific strand types:
+
+- `t"2012-05-26T12-00-00Z"` for date-time strands
+- `u'7c3f4ce8-83c4-458f-b1d7-b28352dea93c'` for UUID’s
+- `r"5:00"` for record id strings
+
+## Syntax Error for Block-RecordId Object ambiguity
+
+The following statement is ambiguous:
+
+```sql
+{ a:b.c }
+```
+
+This can be parsed either as an object with a field `a` and value `b.c`, or as a block statement taking the field `c` of the record `a:b`. Removing this ambiguity completely would require making significant changes to how record IDs are joined (changing the `:` to something else) or requiring that either blocks or objects always be enclosed by parentheses. Both of these changes are quite drastic for a relatively minor ambiguity.
+
+Therefore, I propose that if the parser encounters `{` `Identifier` `:`, it raises a syntax error to notify the user of the ambiguity. The user would then need to either use a record ID strand if they intended to create a block, or make the identifier a strand if they intended to create an object.
+
+# 4. Drawbacks
+
+- Some queries now require more type and syntax than they previously did.
+- These changes break existing code.
+- These new changes could be more difficuly to learn initially as it introduces more syntax rules to remember.
+
+# 5. Alternatives
+
+## Limited reserved word list
+
+Instead of location specific keywords we could instead specify a limited list of keywords which are disallowed. Keywords which can't start a statement like `EVENT` or `TABLE` would still be allowed, but `USE` would be disallowed everywhere.
+
+This would probably be easier to communicate then location specific keywords
+
+# 6. Potential Impacts
+
+The proposed changes are breaking changes. They will probably break existing code, if a limited set. As we have guaranteed stability it introducing these changes as the default would require releasing a new major version. Therefore I propose that the new parser will implement this syntax and introduce the new parser as a new experimental feature which in the future will possibly become the default.
+
+# 7. Unresolved Questions
+
+## What do we change the KNN operator to?
+
+The `knn<3>` operator seems a bit verbose and rather complex looking for an operator. 
+Is this the right syntax.
+
+## How do we differentiate an plain object from a geometry object?
```

Comment:
I am very curious about this. I am not too sure what is currently considered a plain object and a geometry object. Are there any examples?

Reply:
Ah, I see a Geometry example at the bottom of page. 

Is there potential that one may want to do topological analysis on plain objects like vectors or matrices, because then maybe with Geometry as a supertype one can share topology methods with plain objects?

Reply:
Geometry objects are objects which have a very specific layout. The following query returns one kind of geometry object.
```sql
RETURN { 
    type: 'LineString',
    coordinates: [1,2,3,4]
}
```
As you can see it looks exactly like an normal object but if you change even a small part like for example:
```sql
RETURN { 
    type: 'LinesString',
    coordinates: [1,2,3,4]
}
```
(`Lines` instead of `Line`) it is suddenly a plain object and no longer a geometry.

Reply:
I shall add this example to the rfc

Reply:
I think considering making the plain object the supertype may also be an option: one may be able to draw some tangents from geometry kernels and languages like apt for example: https://en.wikipedia.org/wiki/APT_(programming_language). 

At the end of the day Geometry is comprised of plain objects.

Reply:
Are there any semantic differences between geometry objects and plain objects (like operations)? 
If not, then maybe the best thing to do is to treat geometries like plain objects by default and allow for defining the field as otherwise.
This would mean that an application like Surrealist can get information about the table to know how to display those fields.
We can even copy the `record<table>` syntax as `object<schema>`. There is even the potential to allow for user-defined schemas to validate against using a URI (maybe like `object<"http://example.org/bibliography.schema.json">`).

So for a geometry object this might look like:
`DEFINE FIELD position ON TABLE location_history TYPE object<geometry>;`

Reply:
Oh, guess there are operators specialized for geometry objects. Maybe it would be better to move these operators to functions.
For example:
```sql
(-0.118092, 51.509865) INSIDE {
    type: "Polygon",
    coordinates: [[
        [-0.38314819, 51.37692386], [0.1785278, 51.37692386],
        [0.1785278, 51.61460570], [-0.38314819, 51.61460570],
        [-0.38314819, 51.37692386]
    ]]
};
```
Could be:
```sql
geometry::inside((-0.118092, 51.509865), {
    type: "Polygon",
    coordinates: [[
        [-0.38314819, 51.37692386], [0.1785278, 51.37692386],
        [0.1785278, 51.61460570], [-0.38314819, 51.61460570],
        [-0.38314819, 51.37692386]
    ]]
});
```
Then the `INSIDE` operator could be generalized to plain objects. For example:
```sql
{ "foo": "bar" } INSIDE { "foo: "bar", "biz": "fiz" }
```